bt-core/src/main/java/bt/data/digest/JavaSecurityDigester.java

```diff
@@ -25,15 +25,15 @@
 import java.security.NoSuchAlgorithmException;
 import java.util.Arrays;
 
+/**
+ * A class to produce arbitrary digests for {@link DataRange} or {@link Range}. Uses java security libraries to produce the digests
+ */
 public class JavaSecurityDigester implements Digester {
 
     private final String algorithm;
     private final int step;
-    private final SoftThreadLocal<byte[]> bufferTL;
```

Comment:
What's wrong with SoftThreadLocal? This was added not so long ago

Reply:
I've had bad experiences with SoftReferences and avoid them if possible. If having a smaller buffer yields reasonable performance, I think SoftReferences are be unnecessary. If they are necessary, perhaps it's worth considering a buffer pool.

Not all of the below points apply to SoftReference's usage here, but it's why I try to avoid them:

SoftReferences can cause GC to slow down significantly: https://stackoverflow.com/questions/23101817/g1gc-remark-phase-is-taking-too-long
There are open bugs for the JVM not freeing the memory when needed in larger heaps: https://bugs.openjdk.java.net/browse/JDK-8014364
The configuration when this goes wrong is awful - see documentation on `-XX:SoftRefLRUPolicyMSPerMB=` https://docs.oracle.com/en/java/javase/11/tools/java.html#GUID-3B1CE181-CD30-4178-9602-230B800D4FAE 

Also, their implementation isn't the same across JVMs. OpenJ9 has a completely different heuristic for when to SoftReferences.

Reply:
Let's try to summon the author of this code. @wizzardo, if you don't mind me asking – in [99723df5 added thread-local buffer](https://github.com/atomashpolskiy/bt/commit/99723df53a3240256ca3a92bc70265e5c02bc4a7), were you solving a real problem or making a speculative improvement?

Reply:
yep, as I said in PR https://github.com/atomashpolskiy/bt/pull/157 it drastically reduces allocations. And I never had any issues with soft-references, but also didn't abuse them, couple of soft-references here and there..

Reply:
later I revisited these changes and reduced size of buffer, it doesn't need to process whole chunk at once. Currently I have too many local changes to create a pr, and no time to organize them =(
```java
/*
 * Copyright (c) 2016—2017 Andrei Tomashpolskiy and individual contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package bt.data.digest;

import bt.BtException;
import bt.data.DataRange;
import bt.data.range.Range;

import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class JavaSecurityDigester implements Digester {

    private final String algorithm;
    private final SoftThreadLocal<byte[]> bufferTL;
    private final SoftThreadLocal<MessageDigest> digestTL = new SoftThreadLocal<>(() -> createDigest(), messageDigest -> {
        messageDigest.reset();
        return messageDigest;
    });

    public JavaSecurityDigester(String algorithm) {
        try {
            // verify that implementation for the algorithm exists
            MessageDigest.getInstance(algorithm);
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalArgumentException("No such algorithm: " + algorithm, e);
        }
        this.algorithm = algorithm;
        bufferTL = new SoftThreadLocal<>(() -> new byte[16 * 1024]);
    }

    public String getAlgorithm() {
        return algorithm;
    }

    @Override
    public byte[] digest(DataRange data) {
        MessageDigest digest = digestTL.getValue();

        data.visitUnits((unit, off, lim) -> {
            long remaining = lim - off;
            if (remaining > Integer.MAX_VALUE) {
                throw new BtException("Too much data -- can't read to buffer");
            }
            byte[] bytes = bufferTL.getValue();
            do {
                ByteBuffer wrap = ByteBuffer.wrap(bytes);
                if (remaining < wrap.capacity()) {
                    wrap.limit((int) remaining);
                }
                int read = unit.readBlock(wrap, off);
                if (read == -1) {
                    // end of data, terminate
                    return false;
                }
                digest.update(bytes, 0, read);
                remaining -= read;
                off += read;
            } while (remaining > 0);

            return true;
        });

        return digest.digest();
    }

    @Override
    public byte[] digestForced(DataRange data) {
        MessageDigest digest = digestTL.getValue();

        data.visitUnits((unit, off, lim) -> {
            long remaining = lim - off;
            if (remaining > Integer.MAX_VALUE) {
                throw new BtException("Too much data -- can't read to buffer");
            }
            byte[] bytes = bufferTL.getValue();
            ByteBuffer buffer = ByteBuffer.wrap(bytes);
            do {
                if (remaining < buffer.capacity()) {
                    buffer.limit((int) remaining);
                }
                int read = buffer.limit();
                unit.readBlockFully(buffer, off);
                if (buffer.hasRemaining()) {
                    throw new IllegalStateException("Failed to read data fully: " + buffer.remaining() + " bytes remaining");
                }
                digest.update(bytes,0, buffer.flip().limit());
                remaining -= read;
                off += read;
            } while (remaining > 0);

            return true;
        });

        return digest.digest();
    }

    @Override
    public byte[] digest(Range<?> data) {
        MessageDigest digest = digestTL.getValue();
        digest.update(data.getBytes());
        return digest.digest();
    }

    private MessageDigest createDigest() {
        try {
            return MessageDigest.getInstance(algorithm);
        } catch (NoSuchAlgorithmException e) {
            // not going to happen
            throw new BtException("Unexpected error", e);
        }
    }
}

```

Reply:
and if you don't really like soft-references, we can use normal thread-local variables here, soft-reference was needed for huge 8-16mb buffers, if it's just 16kb we can just keep it in memory

Reply:
There's no reason why this buffer ever should be bigger than the a torrent's piece size.

I did more testing - my laptop can digest about 700-750 MB/s with a 16 KiB buffer, and enlarging it doesn't help performance. This seems to roughly be the limit of my laptop's NVMe SSD.  However I've gotten a wide range of benchmark though in my testing, which I suspect is CPU and SSD throttling due to heat. @atomashpolskiy - let me know what your benchmarks show.

I checked in what threads verification is done with this buffer. It's not that many places.

1) DefaultDataWorker's executor (single thread). Keeping this buffer here is what we want, regardless of the size.
2) TorrentMaker - this isn't relevant because the digester becomes garbage after the torrent is created, and any ThreadLocal variables are soon GC'd
3) DefaultChunkVerifier.verifyChunks. Runs in either a ForkJoinPool.commonPool(), a new ForkJoinPool(), or the thread itself. I changed the code for this to make a copy of the Digester so after the verification is done, the buffer allocated by these threads can be GC'd even if the thread exists afterward.